File: profile-service/logger/detail.go
package logger

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"runtime"
	"time"

	"github.com/sing3demons/profile-service/constants"
)

type DetailLog interface {
	IsRawDataEnabled() bool
	AddInputRequest(node, cmd, invoke string, rawData, data interface{})
	AddOutputRequest(node, cmd, invoke string, rawData, data interface{})
	End()
	AddInputResponse(node, cmd, invoke string, rawData, data interface{}, protocol, protocolMethod string)
	AddOutputResponse(node, cmd, invoke string, rawData, data interface{})
	AutoEnd() bool
}

func NewDetailLog(req *http.Request, initInvoke, scenario, identity string, conf LogConfig) DetailLog {
	session := req.Context().Value(constants.Session)

	host, _ := os.Hostname()
	data := &detailLog{
		LogType:       "Detail",
		Host:          host,
		AppName:       conf.ProjectName,
		Instance:      getInstance(),
		Session:       fmt.Sprintf("%v", session),
		InitInvoke:    initInvoke,
		Scenario:      scenario,
		Identity:      identity,
		Input:         []InputOutputLog{},
		Output:        []InputOutputLog{},
		conf:          conf.Detail,
		startTimeDate: time.Now(),
		timeCounter:   make(map[string]time.Time),
		req:           req,
	}

	return data
}

func getInstance() *string {
	instance, err := os.Hostname()
	if err != nil {
		instance = fmt.Sprintf("%d", os.Getpid())
	}
	if instance == "" {
		instance = fmt.Sprintf("%d", os.Getpid())
	}
	return &instance
}

func (dl *detailLog) IsRawDataEnabled() bool {
	return dl.conf.RawData
}

func (dl *detailLog) AddInputRequest(node, cmd, invoke string, rawData, data interface{}) {
	if rawData != nil {
		if _, ok := rawData.(string); !ok {
			rawData = ToJson(rawData)
		}
	}
	dl.addInput(&logEvent{
		node:           node,
		cmd:            cmd,
		invoke:         invoke,
		logType:        "req",
		rawData:        rawData,
		data:           data,
		protocol:       dl.req.Proto,
		protocolMethod: dl.req.Method,
	})
}

func (dl *detailLog) AddInputResponse(node, cmd, invoke string, rawData, data interface{}, protocol, protocolMethod string) {
	resTime := time.Now().Format(time.RFC3339)
	if rawData != nil {
		if _, ok := rawData.(string); !ok {
			rawData = ToJson(rawData)
		}
	}
	dl.addInput(&logEvent{
		node:           node,
		cmd:            cmd,
		invoke:         invoke,
		logType:        "res",
		rawData:        rawData,
		data:           ToStruct(data),
		resTime:        resTime,
		protocol:       protocol,
		protocolMethod: protocolMethod,
	})
}

func (dl *detailLog) AddOutputResponse(node, cmd, invoke string, rawData, data interface{}) {
	if rawData != nil {
		if _, ok := rawData.(string); !ok {
			// rawData = fmt.Sprintf("%v", rawData)
			rawData = ToJson(rawData)
		}
	}
	dl.AddOutput(logEvent{
		node:    node,
		cmd:     cmd,
		invoke:  invoke,
		logType: "res",
		rawData: rawData,
		data:    ToStruct(data),
	})
	// dl.End()
}

func (dl *detailLog) addInput(input *logEvent) {
	dl.mu.Lock()
	defer dl.mu.Unlock()
	now := time.Now()
	if dl.startTimeDate.IsZero() {
		dl.startTimeDate = now
	}

	var resTimeString string
	if input.resTime != "" {
		resTimeString = input.resTime
	} else if input.logType == "res" {
		if startTime, exists := dl.timeCounter[input.invoke]; exists {
			duration := time.Since(startTime).Milliseconds()
			resTimeString = fmt.Sprintf("%d ms", duration)
			delete(dl.timeCounter, input.invoke)
		}
	}

	protocolValue := dl.buildValueProtocol(&input.protocol, &input.protocolMethod)
	inputLog := InputOutputLog{
		Invoke:   input.invoke,
		Event:    fmt.Sprintf("%s.%s", input.node, input.cmd),
		Protocol: protocolValue,
		Type:     input.logType,
		RawData:  dl.isRawDataEnabledIf(input.rawData),
		Data:     input.data,
		ResTime:  &resTimeString,
	}
	dl.Input = append(dl.Input, inputLog)
}

func (dl *detailLog) AddOutputRequest(node, cmd, invoke string, rawData, data interface{}) {
	if rawData != nil {
		if _, ok := rawData.(string); !ok {
			rawData = ToJson(rawData)
		}
	}
	dl.AddOutput(logEvent{
		node:           node,
		cmd:            cmd,
		invoke:         invoke,
		logType:        "rep",
		rawData:        rawData,
		data:           ToStruct(data),
		protocol:       dl.req.Proto,
		protocolMethod: dl.req.Method,
	})
}

func (dl *detailLog) AddOutput(out logEvent) {
	dl.mu.Lock()
	defer dl.mu.Unlock()
	now := time.Now()
	if out.invoke != "" && out.logType != "res" {
		dl.timeCounter[out.invoke] = now
	}

	protocolValue := dl.buildValueProtocol(&out.protocol, &out.protocolMethod)
	outputLog := InputOutputLog{
		Invoke:   out.invoke,
		Event:    fmt.Sprintf("%s.%s", out.node, out.cmd),
		Protocol: protocolValue,
		Type:     out.logType,
		RawData:  dl.isRawDataEnabledIf(out.rawData),
		Data:     out.data,
	}
	dl.Output = append(dl.Output, outputLog)
}

func (dl *detailLog) End() {
	if dl.startTimeDate.IsZero() {
		log.Fatal("end() called without any input/output")
	}

	processingTime := fmt.Sprintf("%d ms", time.Since(dl.startTimeDate).Milliseconds())
	dl.ProcessingTime = &processingTime

	inputTimeStamp := dl.formatTime(dl.inputTime)
	dl.InputTimeStamp = inputTimeStamp

	outputTimeStamp := dl.formatTime(dl.outputTime)
	dl.OutputTimeStamp = outputTimeStamp

	logDetail, _ := json.Marshal(dl)
	if dl.conf.LogConsole {
		os.Stdout.Write(logDetail)
		os.Stdout.Write([]byte(endOfLine()))
	}

	if dl.conf.LogFile {
		dl.conf.LogDetail.Info(string(logDetail))
	}

	dl.clear()
}

func (dl *detailLog) buildValueProtocol(protocol, method *string) *string {
	if protocol == nil {
		return nil
	}
	result := *protocol
	if method != nil {
		result += "." + *method
	}
	return &result
}

func (dl *detailLog) AutoEnd() bool {
	if dl.startTimeDate.IsZero() {
		return false
	}
	if len(dl.Input) == 0 && len(dl.Output) == 0 {
		return false
	}

	dl.End()
	return true
}

func (dl *detailLog) isRawDataEnabledIf(rawData interface{}) interface{} {
	if dl.conf.RawData {
		return rawData
	}
	return nil
}

func (dl *detailLog) formatTime(t *time.Time) *string {
	if t == nil {
		return nil
	}
	ts := t.Format(time.RFC3339)
	return &ts
}

func endOfLine() string {
	if runtime.GOOS == "windows" {
		return "\r\n"
	}
	return "\n"
}

func (dl *detailLog) clear() {
	dl.ProcessingTime = nil
	dl.InputTimeStamp = nil
	dl.OutputTimeStamp = nil
	dl.Input = nil
	dl.Output = nil
	dl.startTimeDate = time.Time{}
}

func ToJson(data interface{}) string {
	jsonData, err := json.Marshal(data)
	if err != nil {
		return fmt.Sprintf("%v", data)
	}
	return string(jsonData)
}

// convert struct to json
func ToStruct(data interface{}) (result interface{}) {
	jsonData, err := json.Marshal(data)
	if err != nil {
		return data
	}

	err = json.Unmarshal(jsonData, &result)
	if err != nil {
		return data
	}
	return result
}

File: auth-service/src/lib/logger/detail.ts
import * as os from 'os'
import dayjs from 'dayjs'
import { confLog, writeLogFile } from './logger.js'
import { IDetailLog, InputOutput, LogConfig, LogData, MaskingConfig } from './types.js'

export default class DetailLog {
  public startTimeDate: Date | null = null
  private inputTime: Date | null = null
  private outputTime: Date | null = null
  private timeCounter: { [key: string]: Date } = {}
  public detailLog: IDetailLog
  private readonly conf: LogConfig = confLog
  constructor(session: string, initInvoke?: string, scenario?: string, identity?: string) {
    this.detailLog = {
      LogType: 'Detail',
      Host: os.hostname(),
      AppName: this.conf.projectName ?? '',
      Instance: process.pid,
      Session: session,
      InitInvoke: initInvoke ?? this.conf.projectName + `_${dayjs().format('yyyymmddHHMMss')}`,
      Scenario: scenario ?? '',
      InputTimeStamp: null,
      Input: [],
      OutputTimeStamp: null,
      Output: [],
      ProcessingTime: null,
    }
  }

  public New(scenario: string) {
    this.detailLog.Input.length = 0
    this.detailLog.Scenario = scenario
    return this
  }

  isRawDataEnabled(): boolean {
    return this.conf.detail.rawData === true
  }
  private readonly maskingConfig: MaskingConfig = {
    highlight: ['email', 'phone'],
    mark: ['password', 'pin', 'otp', 'token', 'secret', 'api-key'],
  }

  private maskSensitiveData<T>(body: T): T {
    if (!body) {
      return body;
    }

    if (typeof body === 'string') {
      return this.maskStringData(body) as T;
    }

    if (typeof body === 'object') {
      return this.maskObjectData(body) as T;
    }

    return body;
  }

  private maskStringData(body: string): string {
    try {
      const parsedBody = JSON.parse(body);
      const data = JSON.stringify(this.maskSensitiveData(parsedBody));
      return data;
    } catch (error) {
      return body;
    }
  }

  private maskObjectData<T>(body: T): T {
    const data = JSON.parse(JSON.stringify(body)); // Clone the object to avoid modifying the original data
    const maskedData: any = Array.isArray(data) ? [] : {};

    for (const key in data) {
      if (data.hasOwnProperty(key)) {
        maskedData[key] = this.maskField(key, data[key]);
      }
    }

    return maskedData;
  }

  private maskField(key: string, value: any): any {
    if (this.maskingConfig?.mark?.includes(key)) {
      return '******'; // Fully mask fields in 'mark'
    }

    if (this.maskingConfig?.highlight?.includes(key)) {
      return this.applyPartialMask(value); // Partially mask fields in 'highlight'
    }

    if (typeof value === 'object') {
      return this.maskSensitiveData(value); // Recursively mask nested objects
    }

    return value;
  }
  private applyPartialMask(value: string): string {
    const rex = new RegExp(/^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/)
    if (rex.test(value)) {
      let [first, second] = value.split('@')
      if (!first) {
        return ''
      }
      if (first.length > 2) {
        const mask = first.substring(3, first.length)
        const notMask = first.substring(0, 3)
        first = notMask + '*'.repeat(mask.length)
      } else {
        first = first.replace(first.substring(1, first.length), '*'.repeat(first.length - 1))
      }
      return `${first}@${second}`
    }

    return value.length > 2 ? value.substring(0, 2) + '*'.repeat(value.length - 2) : value
  }

  private readonly validateData = (data: any): boolean => {
    const { headers, params, query, body } = data

    if (headers) return true

    if (params || typeof params === 'object') {
      return true
    }

    if (query || typeof query === 'object') {
      return true
    }

    if (body || typeof body === 'object') {
      return true
    }

    return false
  }

  addInputRequest(node: string, cmd: string, invoke: string, req: any): void {
    const data = {
      headers: {
        device: req?.deviceInfo,
        session: req?.session,
        userId: req?.userId,
        host: req?.hostname,
        ip: req?.ip,
      },
      params: req?.params,
      query: req?.query,
      body: req?.body,
    }

    if (invoke === '') {
      invoke = this.detailLog.InitInvoke
    }

    if (!this.validateData(data)) {
      this.addInput(node, cmd, invoke, 'req', '', req, undefined, req?.protocol, req?.method)
    } else {
      this.addInput(node, cmd, invoke, 'req', '', data, undefined, req?.protocol, req?.method)
    }
  }

  addInputResponseError(node: string, cmd: string, invoke: string, rawData?: string): void {
    this.addInput(node, cmd, invoke, 'res_error', rawData)
  }

  addInputResponse(node: string, cmd: string, invoke: string, rawData: string, data: LogData, resTime?: number): void {
    this.addInput(node, cmd, invoke, 'res', rawData, data, resTime)
  }

  private addInput(
    node: string,
    cmd: string,
    invoke: string,
    type: string,
    rawData?: string,
    data?: LogData,
    resTime?: number | string,
    protocol?: string,
    protocolMethod?: string
  ): void {
    this.inputTime = new Date()

    if (!this.startTimeDate) {
      this.startTimeDate = this.inputTime
    }

    if (rawData) {
      rawData = this.maskSensitiveData(rawData)
    }

    if (data) {
      data = this.maskSensitiveData(data)
    }

    if (typeof resTime === 'number') {
      resTime = resTime + ' ms'
      delete this.timeCounter[invoke]
    } else if (type.startsWith('res')) {
      if (this.timeCounter[invoke]) {
        resTime = this.inputTime.getTime() - this.timeCounter[invoke].getTime()
        resTime = resTime + ' ms'
        delete this.timeCounter[invoke]
      }
    }

    const input: InputOutput = {
      Invoke: invoke,
      Event: `${node}.${cmd}`,
      Protocol: type === 'req' ? this._buildValueProtocol(protocol, protocolMethod) : undefined,
      Type: type,
      RawData: this.conf.detail.rawData === true ? rawData : undefined,
      Data: data,
      ResTime: resTime,
    }

    this.detailLog.Input.push(input)
  }

  addOutputRequest(
    node: string,
    cmd: string,
    invoke: string,
    rawData: string,
    data: LogData,
    protocol?: string,
    protocolMethod?: string
  ) {
    this.addOutput(node, cmd, invoke, 'req', rawData, data, protocol, protocolMethod)
    return this
  }

  addOutputResponse(node: string, cmd: string, invoke: string, rawData: string, data: LogData) {
    this.addOutput(node, cmd, invoke, 'res', rawData, data)
    return this
  }

  addOutputRequestRetry(
    node: string,
    cmd: string,
    invoke: string,
    rawData: string,
    data: LogData,
    total: number,
    maxCount: number
  ): void {
    this.addOutput(node, cmd, invoke, `req_retry_${total}/${maxCount}`, rawData, data)
  }

  private addOutput(
    node: string,
    cmd: string,
    invoke: string,
    type: string,
    rawData: string,
    data: LogData,
    protocol?: string,
    protocolMethod?: string
  ): void {
    this.outputTime = new Date()

    if (invoke && type !== 'res') {
      this.timeCounter[invoke] = this.outputTime
    }

    if (rawData) {
      rawData = this.maskSensitiveData(rawData)
    }

    if (data) {
      data = this.maskSensitiveData(data)
    }

    const output: InputOutput = {
      Invoke: invoke,
      Event: `${node}.${cmd}`,
      Protocol: type === 'req' ? this._buildValueProtocol(protocol, protocolMethod) : undefined,
      Type: type,
      RawData: this.conf.detail.rawData === true ? rawData : undefined,
      Data: data,
    }

    this.detailLog.Output.push(output)
  }

  public isEnd(): boolean {
    return this.startTimeDate === null
  }

  end(): void {
    if (!this.startTimeDate) throw new Error('detailLog call "end()", twice')

    this.detailLog.ProcessingTime = new Date().getTime() - this.startTimeDate.getTime() + ' ms'
    this.detailLog.InputTimeStamp = this.inputTime && dayjs(this.inputTime).format()

    if (this.outputTime) {
      this.detailLog.OutputTimeStamp = dayjs(this.outputTime).format()
    } else {
      this.detailLog.Output.length = 0
      this.detailLog.OutputTimeStamp = null
    }

    // Writing the log (simulating)
    const log = {
      systemTimestamp: this.detailLog.InputTimeStamp,
      logType: 'Detail',
      logLevel: 'INFO',
      namespace: this.conf.namespace,
      containerId: os.hostname(),
      applicationName: this.detailLog.AppName,
      detail: this.detailLog,
    }

    if (this.conf.detail.console) {
      process.stdout.write(JSON.stringify(log) + os.EOL)
    }

    if (this.conf.detail.file) {
      writeLogFile('dtl', JSON.stringify(log))
    }

    this._clr()
  }

  private _clr(): void {
    this.detailLog.ProcessingTime = null
    this.detailLog.InputTimeStamp = null
    this.detailLog.OutputTimeStamp = null
    this.detailLog.Input.length = 0
    this.detailLog.Output.length = 0
    this.outputTime = null
    this.startTimeDate = null
  }

  private _buildValueProtocol(protocol: string | undefined, protocolMethod: string | undefined): string | undefined {
    let v = undefined
    if (protocol) {
      v = protocol.toLowerCase()
      if (protocolMethod) {
        v = `${v}.${protocolMethod.toLowerCase()}`
      }
    }
    return v
  }
}
